### DISCUSSION TITLE
What do you think of this Alpine component pattern?

### DISCUSSION THREAD

**Discussion by curvedinf** [alpinejs/alpine#3777] [4. General]
Hey folks, I just wanted to discuss component patterns for Alpine. What patterns are you using?

Here's the pattern I've settled on:
https://medium.com/@djangoist/a-modern-component-pattern-for-alpinejs-integrated-with-server-side-rendering-examples-using-d6eca21dcd19

If you have a moment to read the article, I'd like to hear considerations you have or suggested improvements. Thank you!

**Comment by ekwoka** [comment]
You can pass arguments to the function called in x-data.

That's the more preferred for simple data like shown here.

`x-data="button('display text')"`

That's more common than a prop.

If you need a larger amount of data that might not be serializable in a safe way to pass in an html attribute, we normally do it in a script tag as the first child.

```html
<div x-data="complexThing">
  <script type="application/JSON">
     {{ big_object|json }}
   </script>
   ...
</div>
```

Then you can just parse it from there.


Mostly, Django is a huge PITA. Alpine makes dealing with these things pretty okay, but Django makes things pretty bad that it's still hard to recommend Django + Alpine.


One thing also somewhat missed in the article is that with htmx, if you're boosting the navigation, using Alpine:init events will totally fall apart.

**Reply by curvedinf** [reply]
Good to know about passing data into the x-data function. Why is Django a PITA? For the medium sized apps I work on I haven't found any other framework that makes life easier. @ekwoka

**Reply by ekwoka** [reply]
Templates a poor excuse for components.

The ORM system is really bad and wasteful.

Everything is designed to just be classes extending classes extending classes extending classes, resulting is just way too much unclear magic.

The form system is pretty close to useless with way too many gotchas that don't have simple solutions.

Even NextJS,  which isn't even the best typescript option, is a much easier tool to work with, especially something like T3.

That's not even counting the fact python is so poorly designed and poor performing.

I work in Django, so this isn't like I tried doing one thing once.

Can't really trust the opinions of anyone that thinks Django is decent, much less good.

But that's mostly neither here nor there.

Passing props to the data function is the solution you want 😁

**Reply by curvedinf** [reply]
I am curious since you're so opinionated, what project has a good ORM if Django's is bad to you?

**Reply by ekwoka** [reply]
Prisma is a decent ORM.

Supabase has a decent one as well.

I think a big part of it should be that it requires explicit making of queries, not just doing background queries, especially BLOCKING queries like Django does.

It should be about providing e2e type safety and building safe queries.

**Reply by reedjones** [reply]
"That's not even counting the fact python is so poorly designed and poor performing." - Python has [one of the best language design philosophies ever](https://peps.python.org/pep-0020/ ). 

Nodejs design philosophy = "blazing fast scale big"

**Reply by ekwoka** [reply]
> "That's not even counting the fact python is so poorly designed and poor performing." - Python has [one of the best language design philosophies ever](https://peps.python.org/pep-0020/).

Nice in concept. Not true in reality. That's a nice philosophy that doesn't describe python.

There is no excuse for any of the `__name__` things or how they are used and abused.

JS isn't perfect either, but most of the issues are avoided in real code, not leaned into.
